API endpoint to create new VLANs in VLANGroup. (#6273)
nautobot/nautobot: nautobot/docs/user-guide/core-data-model/ipam/vlangroup.md

VLAN groups can be used to organize VLANs within Nautobot. Each group may optionally be assigned to a specific location, but a group cannot belong to multiple locations.

Groups can also be used to enforce uniqueness: Each VLAN within a group must have a unique ID and name. VLANs which are not assigned to a group may have overlapping names and IDs (including VLANs which belong to a common location). For example, you can create two VLANs with ID 123, but they cannot both be assigned to the same group.

## VLAN Group Ranges

+/- 2.4.0

> VLAN Groups contain a mandatory `range` field with a default value of `1-4094` which equals to all VLANs. This field may be used to constrain member VLANs of the group. Only VLANs with VLAN identifiers ("vids") within a `range` value will be allowed to associated with a `VLANGroup`

Reviewer: ```suggestion
VLAN Groups contain a mandatory `range` field with a default value of `1-4094` (permitting all VLANs). This field may be used to constrain the valid member VLANs of the group. VLANs can only be associated with a given VLAN Group if their VLAN identifiers ("vids") fall within the specified `range`.
```

Range value accepts commas and dashes, with examples as follows:

* 1
* 1-5
* 1-5,10
* 1-5,10-15
* 1-5,10-15,16,17,18,19,20

Values between dashes will also be expanded into a list of VLANs.

## Creating new VLANs in a VLANGroup programatically

+/- 2.4.0

> VLAN Groups offer new API endpoint to list available and create new VLANs programatically. Nautobot accepts API calls to the following endpoint:

Reviewer: ```suggestion
VLAN Groups offer an API endpoint to list available VIDs and create new VLANs programmatically:
```

`/api/ipam/vlan-groups/<vlan_group_id>/available-vlans/`

> For the `GET` request, API will return a list of all available VLAN IDs available in the given `VLANGroup`.

Reviewer: ```suggestion
A `GET` request to this API endpoint will return a list of all available (unassigned) VLAN IDs available in the given `VLANGroup`.
```

Reviewer 2: Should we keep the language consistent throughout this doc? `VLAN IDs` vs. `VIDs`

> For the `POST` request, API will create requested number of VLANs in a `VLANGroup`.

Reviewer: ```suggestion
A `POST` request to the same API endpoint will create `VLAN` record(s) in the given `VLANGroup`. Note that you must specify the `name` and `status` of each VLAN to be created as a part of this request, but the `vlan_group` and `vid` may be omitted as they can be automatically assigned by this API. See below.
```

### Examples

To create a new `VLANGroup` with a `range` field of: `5-10` (permit VLANs `5, 6, 7, 8, 9, 10`), issue following POST request:

```no-highlight
curl -X POST -H "Authorization: Token $TOKEN" \
-H "Content-Type: application/json" \
-H "Accept: application/json; version=2.0; indent=4" \
http://nautobot/api/ipam/vlan-groups/ \

> --data '[{"name": "Enterprise VLANs", "range": "5-10"}]'

Reviewer 2: Are we missing VLAN status in this example request?

Author: This is a request to the `VLANGroup` and that model does not have a status

```

`VLANGroup` was created as follows:

```json
[
    {
        "id": "c16d5d1d-40d6-4902-b89d-6294e60cb4bf",
        "object_type": "ipam.vlangroup",
        "display": "Enterprise VLANs",
        "url": "http://nautobot/api/ipam/vlan-groups/c16d5d1d-40d6-4902-b89d-6294e60cb4bf/",
        "natural_slug": "enterprise-vlans_c16d",
        "name": "Enterprise VLANs",
        "description": "",
        "range": "5-10",
        "location": null,
        "created": "2024-09-25T09:27:16.343655Z",
        "last_updated": "2024-09-25T09:27:16.343670Z",
        "notes_url": "http://nautobot/api/ipam/vlan-groups/c16d5d1d-40d6-4902-b89d-6294e60cb4bf/notes/",
        "custom_fields": {},
        "computed_fields": {},
        "relationships": {}
    }
]
```

> Then, to list all available `VLANs` within an `Enterprise VLANs` `VLANGroup` issue a GET request:

Reviewer: ```suggestion
Then, to list all available VLAN IDs within the just-created`VLANGroup`, issue a GET request:
```

Reviewer 2: Just a missing space in Glenn's review comment
```suggestion
Then, to list all available VLAN IDs within the just-created `VLANGroup`, issue a GET request:
```

```no-highlight
curl -X GET -H "Authorization: Token $TOKEN" \
-H "Content-Type: application/json" \
-H "Accept: application/json; version=2.0; indent=4" \
http://nautobot/api/ipam/vlan-groups/c16d5d1d-40d6-4902-b89d-6294e60cb4bf/available-vlans/
```

```json
{
    "count": 6,
    "next": null,
    "previous": null,
    "results": [
        5,
        6,
        7,
        8,
        9,
        10

>     ]
> 
> ```

Reviewer: ````suggestion
    ]
}
```
````

> To create two new VLANs, a data payload has to be specified. In the case of first VLAN, its `VLAN ID` will be automatically determined based on `VLANGroups` availability, in the case of second VLAN we manually define `vid` requsting `VLAN` with `id` 8:

Reviewer: Can we elaborate more on the data payload, i.e. mandatory `name` and `status`, optional `vid`, `tenant`, `role`, etc.?

        "display": "Enterprise VLANs",
        "url": "http://nautobot/api/ipam/vlan-groups/c16d5d1d-40d6-4902-b89d-6294e60cb4bf/",
        "natural_slug": "enterprise-vlans_c16d",
        "name": "Enterprise VLANs",
        "description": "",
        "range": "5-10",
        "location": null,
        "created": "2024-09-25T09:27:16.343655Z",
        "last_updated": "2024-09-25T09:27:16.343670Z",
        "notes_url": "http://nautobot/api/ipam/vlan-groups/c16d5d1d-40d6-4902-b89d-6294e60cb4bf/notes/",
        "custom_fields": {},
        "computed_fields": {},
        "relationships": {}
    }
]
```

### Listing available VLANs example

Then, to list all available VLAN IDs within the just-created `VLANGroup`, issue a GET request:

```no-highlight
curl -X GET -H "Authorization: Token $TOKEN" \
-H "Content-Type: application/json" \
-H "Accept: application/json; version=2.0; indent=4" \
http://nautobot/api/ipam/vlan-groups/c16d5d1d-40d6-4902-b89d-6294e60cb4bf/available-vlans/
```

```json
{
    "count": 6,
    "next": null,
    "previous": null,
    "results": [
        5,
        6,
        7,
        8,
        9,
        10
    ]
}
```

### Creating new VLANs example

> `/api/ipam/vlan-groups/<vlan_group_id>/available-vlans/` API endpoint accepts data similar to the `VLAN` API endpoint.
> However, there are some characteristic differences while comparing to the `VLAN` endpoint:
> 
> Notes about `/available-vlans/`:
> 
> * accepts `GET` requests and returns list of available VLAN IDs within a VLAN Group
> * accepts `POST` requests
>     * accepts a list of dictionaries to create multiple VLANs (single dictionary is also permitted), returns a list of VLAN objects
>     * accepts a single dictionary to create 1 VLAN, returns a single dictionary describing a VLAN
>     * fields `name` and `status` of `VLAN` object are mandatory
>     * field `vid` is optional and might be specified to override auto-allocation for given VLAN
>     * field `locations` is not supported
>     * other `VLAN` model fields are permitted and optional
>     * will not accept an optional `vlan_group` attribute different than specified in request's URL
>     * will not accept a request with the same VLAN IDs specified multiple times
> 
> To create two new VLANs, a data payload has to be specified. In the case of the first VLAN, its `VLAN ID` will be automatically determined based on `VLANGroups` availability, in the case of second VLAN we manually define `vid` requsting `VLAN` with `id` 8:

Reviewer: ```suggestion
The `/api/ipam/vlan-groups/<vlan_group_id>/available-vlans/` API endpoint accepts `POST` requests to create a new VLAN record, similar to the `/api/ipam/vlans/` API endpoint. However, there are some characteristic differences in the behavior of `/available-vlans/` with regard to its input and output:

* Accepts a single dictionary as input to create 1 VLAN, in which case it returns a single dictionary describing the created VLAN
* Accepts a list of dictionaries as input to create one or more VLANs, in which case it returns a list of VLAN objects
* The fields `name` and `status` of are mandatory inputs for each VLAN to be created
* The field `vid` is optional and might be specified to override auto-allocation for a given VLAN
    * Will not accept a request with the same `vid` value specified multiple times
* Will not accept a `vlan_group` attribute different than that already specified in the request's URL
* Other `VLAN` model fields are permitted and optional
* As with the `/vlans/` endpoint, specifying `locations` in the POST request is not presently supported; you can make a subsequent call(s) to the `/api/ipam/vlan-location-assignments/` endpoint to update the many-to-many association between the created VLAN(s) and desired Location(s)

To create two new VLANs, a data payload has to be specified. In the case of the first VLAN, its `VLAN ID` will be automatically determined based on `VLANGroups` availability, in the case of the second VLAN we explicitly request `vid` 8:
```
